### PR TITLE
Merge dev into master

### PR DESCRIPTION
**Pull request by bleaker0x**
Notable changes:

- use MiniRpc for RPC connections by default 
- Reduce network time spent during token swap requests (#51)
- Cronos and Metis chains integration (#52)
- Fix bridge transactions initiated as gas token transfers erroring out (#41, #58)
- Improve and/or fix logic in bridge transaction builder functions
- Add more test cases for bridge transaction logic
- Implement various tests and test cases to bring overall code coverage up
- Enable and use parallel testing
- Update frontend example for latest code in dev
- Integrate BridgeConfigV3 contract, replacing BridgeConfig contract. 
- Implement tests for approval and bridge transactions created or executed by Bridge.SynapseBridge

### CODE REVIEW

File: src/entities.ts [hunk 1]
```diff
@@ -39,53 +42,53 @@ export const newL2BridgeZapInstance = (params: {
...
+            rpcProviderForNetwork(ChainId.ETH)
+        )
+
+    export const bridgeConfigV3 = (): BridgeConfigV3Contract =>
+        BridgeConfigV3Factory.connect(
+            bridgeConfigV3Address,
+            rpcProviderForNetwork(ChainId.ETH)
+        )
+
+    export const poolConfig = (): PoolConfigContract =>
```
Comment (aureliusbtc) [inline src/entities.ts:89]: PoolConfig is now merged into BridgeConfigV3 which stores poolconfig: https://github.com/synapsecns/synapse-contracts/blob/162745e7c2e7a5d2012f1b12ea6bc95391ca4c3c/contracts/bridge/BridgeConfigV3.sol#L399

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -33,35 +33,46 @@ jobs:
...
+          BRIDGE_INTERACTIONS_PRIVKEY_ADDRESS: ${{ secrets.BRIDGE_INTERACTIONS_PRIVKEY_ADDRESS }}
 
-      - name: Coverage report
+      - name: Generate Coverage report
         run: yarn -s run coverage:report
         env:
           REPORTS_DIR: ${{ github.workspace }}/coverage
 
-      - name: Coveralls
+      - name: "Coveralls: process coverage report"
```
Comment (trajan0x) [inline .github/workflows/tests.yaml:62]: Do we need quotes here?

File: src/internal/rpcproviders.ts [hunk 3]
```diff
@@ -1,84 +1,85 @@
...
+    [ChainId.MOONBEAM]:  "MOONBEAM_RPC_URI",
+    [ChainId.MOONRIVER]: "MOONRIVER_RPC_URI",
+    [ChainId.ARBITRUM]:  "ARBITRUM_RPC_URI",
+    [ChainId.AVALANCHE]: "AVALANCHE_RPC_URI",
+    [ChainId.AURORA]:    "AURORA_RPC_URI",
+    [ChainId.HARMONY]:   "HARMONY_RPC_URI",
 }
 
-const CHAIN_RPC_URIS: {[c: number]: string} = {
+const CHAIN_RPC_URIS: StringMap = {
```
Comment (trajan0x) [inline src/internal/rpcproviders.ts:27]: Do we want to allow these to be overriden in some way? The .env was cumbersome, but several networks block cors unless you run your own node. Something to consider

**Review by aureliusbtc — APPROVED** [review]

File: test/helpers/index.ts [hunk 4]
```diff
@@ -17,26 +20,24 @@ const testAmounts: string[] = [
     "669", "555",
 ];
 
+// Completely clean privkey with low balances.
```
Comment (trajan0x) [inline test/helpers/index.ts:23]: This needs to use a hardhat provider for tests. Mocking out non-networking tests w/ a private key like this is fine, but this is too flaky for use here

**Review by trajan0x — CHANGES_REQUESTED** [review]

**Comment by trajan0x** [comment]
Good to merge once comments are addressed. Test mocking is probably the biggest item here

File: src/entities.ts (hunk 1 above)
Comment (bleaker0x) [reply src/entities.ts]: Fixed in [e5fe607](https://github.com/synapsecns/sdk/pull/63/commits/e5fe60795630afe9c371ab3fb947aca069f8d7bc)

File: .github/workflows/tests.yaml (hunk 2 above)
Comment (bleaker0x) [reply .github/workflows/tests.yaml:75]: with the colon, yes, but will rename instead

Comment (bleaker0x) [reply .github/workflows/tests.yaml]: Changed in [46f40a8](https://github.com/synapsecns/sdk/pull/63/commits/46f40a8dbcb414b10025f8d2013c2063343c3521)

File: src/internal/rpcproviders.ts (hunk 3 above)
Comment (bleaker0x) [reply src/internal/rpcproviders.ts:27]: If possible, I'd like it to be _slightly_ agnostic to frontend or backend; see the use of `isomorphic-fetch` in [minirpc.ts](https://github.com/synapsecns/sdk/blob/dev/src/internal/minirpc.ts#L1). 

I initially had an implementation for dynamic control of the RPC uris, but ultimately removed it -- putting it back (and rewriting it in a much better way) wouldn't be hard, though.

**Comment by bleaker0x** [comment]
> Good to merge once comments are addressed. Test mocking is probably the biggest item here

@trajan0x I agree with the need for mocking for certain things. However, I do think it's important for the tests to reflect real-world functionality to a decent degree. If you think that not having mock tests is a merge-blocker, then I'll do that, but I do think it's a slightly larger to-do which should be a separate PR.

File: test/helpers/index.ts (hunk 4 above)
Comment (bleaker0x) [reply test/helpers/index.ts:23]: See https://github.com/synapsecns/sdk/pull/63#issuecomment-1053601134

File: src/internal/rpcproviders.ts (hunk 3 above)
Comment (bleaker0x) [reply src/internal/rpcproviders.ts:27]: Regarding CORS, I've tested this code against the frontend example (which would be making requests from localhost) and RPC functions work fine on that front.

**Review by aureliusbtc — APPROVED** [review]

File: src/internal/rpcproviders.ts (hunk 3 above)
Comment (trajan0x) [reply src/internal/rpcproviders.ts:27]: I'm more thinking about the cases where we've had to rapidly switch rpcs (Harmony or more recently boba)

Comment (bleaker0x) [reply src/internal/rpcproviders.ts:27]: I actually came up with a decent implementation for setting RPC provider endpoints individually; I can also write up a function which takes a map of `{ chain id => endpoint }` and sets the endpoint for any chain in the map while leaving unpassed chains untouched

Comment (bleaker0x) [reply src/internal/rpcproviders.ts:27]: Unfortunately, given the nature of a static NPM package, we can't necessarily update RPC endpoints all the time; we can make it very, very easy for users to set the ones being used, and provide a doc page listing a bunch of commonly used and generally working endpoints

Comment (BlazeWasHere) [reply src/internal/rpcproviders.ts:27]: The latter is a better idea, ideally in production you would use your own RPC providers.

Comment (bleaker0x) [reply src/internal/rpcproviders.ts:27]: @BlazeWasHere @trajan0x check out [51a4f87](https://github.com/synapsecns/sdk/pull/63/commits/51a4f87f019f05017975c5e5076dae561bffaa88)

File: test/basic/Basic-test.ts [hunk 5]
```diff
@@ -44,6 +52,38 @@ describe("Basic tests", function(this: Mocha.Suite) {
         )
     })
 
+    describe("Test configureRPCEndpoints", function(this: Mocha.Suite) {
+        const rpcConfig: RPCEndpointsConfig = {
```
Comment (trajan0x) [inline test/basic/Basic-test.ts:56]: Nice

**Pull request merged**